### PR TITLE
docs(blog): add post on stream-batch unification

### PR DESCRIPTION
**Pull request by chloeh13q**
## Description of changes

Add a new blog post on stream-batch unification.

### CODE REVIEW

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 1]
```diff
@@ -0,0 +1,414 @@
+---
+title: "Stream-batch unification through Ibis lens"
```
Comment (chloeh13q) [inline docs/posts/flink-unified-stream-batch/index.qmd:2]: I considered titling this "ELI5: How Ibis unifies streaming and batch"... LOL Perhaps there is a better title

**Comment by chloeh13q** [comment]
`docs_pr` is currently failing because I am linking to the Flink tutorial that hasn't been released somewhere in the blog post

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+author: "Chloe"
+date: 2024-02-08
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:4]: update -- let's either do this Thursday (this week) or Monday -- I want to get the roadmap out, so maybe roadmap Thursday and this Monday? then again maybe it's better to capitalize on the 8.0 launch? then again maybe it's better to wait and re-ignite the streaming fire next week? idk

let's plan on this going out Monday tentatively, roadmap this Thursday

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+tool, and we stumbled upon Ibis as we attempted to build our own translation
+layer that could turn user declarations into relation trees, then optimize and
+deploy the query plan, all while maintaining the underlying infrastructure for
+the user. We considered and prototyped with a number of tools and eventually
+chose Ibis. It had already established a position in the batch world and had
+support for 10+ of the most popular batch engines (at the time). We loved the
+idea of decoupling the user-facing interface from the execution engine, so that
+ users can swap out the execution engine depending on their needs, without
+ having to rewrite code. And, of course, it was open-source. It was everything
+ we dreamed of.
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:22]: some form of this could be good for the testimonials page

also unnecessary spaces on lines 20-22

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+A few months later, [we started introducing Apache Flink as the first streaming
+backend into Ibis](https://github.com/ibis-project/ibis/pull/6408). We saw so
+much more that Ibis can do when it steps outside of batch.
+
+Ibis 8.0 marks the official launch of the first streaming backends in Ibis
+(Apache Flink and RisingWave). This is a very significant milestone in Ibis
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:29]: nit: link to backend pages for Flink and RisingWave

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 5]
```diff
@@ -0,0 +1,414 @@
+---
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:1]: haven't fully read this yet but if it's about batch/streaming unification in general, and not specific to Flink, might want to change the filename to drop the `flink-`

Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:1]: haven't fully read this yet but if it's about batch/streaming unification in general, and not specific to Flink, might want to change the filename to drop the `flink-`

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+Ibis 8.0 marks the official launch of the first streaming backends in Ibis
+(Apache Flink and RisingWave). This is a very significant milestone in Ibis
+development.
+
+You may be wondering: what does this mean? Why is this such a big deal? I will
+be answering these questions in this blog post.
+
+## Ibis combines stream and batch into a single framework beyond version 8.0
+
+Today, Ibis provides support for close to 20 backends, including Dask, DuckDB,
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:37]: 20+ backends

Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:37]: "backends including Dask, ..."

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+
+This means that Ibis is an extremely valuable tool, but it was limited to batch
+workloads. In the case of streaming workloads, where systems are [designed with
+unbounded data in mind](https://www.oreilly.com/radar/the-world-beyond-batch-streaming-101/),
+the batch-oriented Ibis fell short. To deal with an infinite data stream,
+streaming data systems operate with unique concepts such as “event time”,
+“processing time”, “watermark”, etc. All of these were missing from Ibis.
+
+At the same time, streaming systems (Spark Streaming, Apache Flink, RisingWave,
+etc) have been gaining use cases. It drove the development of more mature
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:51]: gaining use cases? or gaining popularity?

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+the dataframe API from backend execution, so that the logic for defining data
+transformations is unaffected by implementation discrepancies across backend
+engines. There is also an ongoing effort to further increase interoperability
+across different languages and systems via a standard query plan intermediate
+representation (IR), using a library called
+[`Substrait`](https://substrait.io/).
+
+What does this actually look like? For example, Ibis allows users to define
+window aggregations using the [`over()`
+method](https://ibis-project.org/reference/expression-tables#ibis.expr.types.groupby.GroupedTable.over).
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:109]: this should be a "local" Quarto link instead of going to the https

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+SELECT [column_list]
+FROM (
+   SELECT [column_list],
+     ROW_NUMBER() OVER ([PARTITION BY col1[, col2...]]
+       ORDER BY time_attr [asc|desc]) AS rownum
+   FROM table_name)
+WHERE rownum = 1
+```
+In a database like Postgres, this could be as simple as
+```sql
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:133]: would add newline here (and a few places below) for consistency -- doesn't really matter for rendering

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+FROM functional_alltypes t0
+```
+And in `pandas`, you would use the method `drop_duplicates()`:
+```python
+df.drop_duplicates()
+```
+
+::: {.callout-note}
+We’re working on supporting deduplication via `distinct()` in Flink backend and
+this feature should be available before the next major release!
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:144]: are you sure? wouldn't publicly promise features in a given release unless we are very sure

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+and a streaming Python DataFrame API and we’re excited about what’s to come
+next. We hope that Ibis can close the gap between batch and streaming in such a
+way that we no longer talk about the two separately, but, rather, as two parts
+of the same paradigm. Streaming naturally lends itself to batch: batch is
+technically just a special case of streaming, where the unbounded data flow
+stops at some point.
+
+Of course, this is only the beginning. There are still technical challenges to
+be solved (e.g., backfill, window computations over large windows, GPU
+acceleration), and we hope we'll have more exciting updates to share with the
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:410]: recommend removing "hope" here, was used in the last paragraph, and we'll definitely have more exciting updates to share!

File: docs/posts/flink-unified-stream-batch/index.qmd [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+community soon!
+
+Check out the new Apache Flink and RisingWave backends and let us know what you
```
Comment (lostmygithubaccount) [inline docs/posts/flink-unified-stream-batch/index.qmd:413]: link to backend pages

**Review by lostmygithubaccount — CHANGES_REQUESTED** [review]
need to re-review later but some initial comments

I'm also going to merge in origin/main ~which might be a mistake...~

File: docs/posts/flink-unified-stream-batch/index.qmd (hunk 10 above)
Comment (chloeh13q) [reply docs/posts/flink-unified-stream-batch/index.qmd:144]: changing it to "soon" 😂

File: docs/posts/unified-stream-batch/index.qmd [hunk 13]
```diff
@@ -0,0 +1,419 @@
+---
+title: "Stream-batch unification through Ibis lens"
```
Comment (lostmygithubaccount) [inline docs/posts/unified-stream-batch/index.qmd:2]: nitpicky, but should this just be "Stream-batch unification through Ibis"? not sure the "lens" at the end is doing much

File: docs/posts/unified-stream-batch/index.qmd [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+author: "Chloe"
```
Comment (lostmygithubaccount) [inline docs/posts/unified-stream-batch/index.qmd:3]: can you add your last name? despite my previous shenanigans we should be consistent

File: docs/posts/unified-stream-batch/index.qmd [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+
+con = ibis.get_backend()
+con.create_table("payments", df)
+```
+
+::: {.callout-note}
+The default execution engine for Ibis is DuckDB.
+:::
+
+This is a series of records of order transactions. At Company Potclay, we have
```
Comment (lostmygithubaccount) [inline docs/posts/unified-stream-batch/index.qmd:236]: might be worth noting this is a hypothetical scenario/fake company

**Review by lostmygithubaccount — CHANGES_REQUESTED** [review]
a few minor requests, thanks for this blog and let's merge today!

**Review by lostmygithubaccount — APPROVED** [review]
thanks!
